Feature/dynamic upstreams

kate-osborn opened the pull request:
Adds support for dynamic upstreams. 

Previously, NKG used the cluster IP of the Service to route traffic to the backend Services specified by HTTPRoutes. With this PR, NKG will use the endpoints of the Pods corresponding to a Service as the upstream servers for a backend Service. 

This change adds the following components:

- EndpointSlice controller. This controller subscribes to EndpointSlice CRUD events and indexes EndpointSlices by their Service owners. This allows us to perform O(1) lookups of EndpointSlices by Service owners.
- Relationship.Capturer. This component is responsible for tracking relationships between Gateway API resources (e.g., HTTPRoutes) and non-Gateway API resources (e.g., Services). This allows us to determine when a Service or EndpointSlice update requires an nginx reload. 
- ServiceResolver. This component replaces the ServiceStore and is responsible for resolving a Service and Port to a list of endpoints. 

This PR also adds upstreams to the nginx config generator. One upstream is generated for each unique and valid Service:Port backend ref. If a backendRef cannot be resolved, an upstream is generated that returns a 502. In addition, a default upstream that returns a 502 is generated to handle the case where a backend ref is invalid. 

Known limitations:

- Traffic cannot be routed to Headless Services that do not have a defined port. While this is a valid Kubernetes Service configuration, we do not currently understand the use case for it. I will do some research and see what I can find out.
- If a user manually creates an EndpointSlice, they will need to populate set the `kubernetes.io/service-name` label, and set the ready condition of the endpoints to true. Otherwise, NKG will fail to resolve the Service endpoints.

pleshakov reviewed (changes requested):
@kate-osborn  please see my review

A few more things:

(1)
Could we update the api compatibility doc? The part about backendRefs in HTTPRoute
> NGINX Kubernetes Gateway will use the IP of the Service as a backend, not the IPs of the corresponding Pods. Watching for Service updates is not supported.

(2) I noticed a bug:

(a) deploy cafe example
(b) confirm everything is deployed and working
(c) change tea svc port from 80 to 7777 in the service manifest
 bug: config doesn’t get regenerated - keeps using the old endpoints.

kate-osborn commented on internal/state/graph_test.go:
```diff
@@ -1075,3 +1128,159 @@ func TestValidateGatewayClass(t *testing.T) {
+		}
+
+		if diff := helpers.Diff(tc.expSvc, svc); diff != "" {
+			t.Errorf("getBackendServiceFromRouteRule() returned incorrect backend service for test case: %q, diff: %+v", tc.msg, diff)
```
Don't need it in this case. I'll remove it.

f5yacobucci commented on internal/events/handler_test.go:
```diff
@@ -55,9 +55,9 @@ var _ = Describe("EventHandler", func() {
 		Expect(fakeGenerator.GenerateCallCount()).Should(Equal(1))
 		Expect(fakeGenerator.GenerateArgsForCall(0)).Should(Equal(expectedConf))
 
-		Expect(fakeNginxFimeMgr.WriteHTTPServersConfigCallCount()).Should(Equal(1))
```
nit: fakeNginxFimeMgr -> fakeNginxFileMgr

kate-osborn commented on internal/state/change_processor.go:
```diff
@@ -75,40 +82,135 @@ func (c *ChangeProcessorImpl) CaptureUpsertChange(obj client.Object) {
+	return false
+}
+
+func (c *ChangeProcessorImpl) updateServicesMap(hr *v1beta1.HTTPRoute) {
```
I reworked the change processor code and added a new component relationship.Capturer that captures relationships between Services and Routes, as well as Endpoints and Services. 

bda6dcf078e9dda1b5386c07dd0579f6e2610262

Let me know what you think.

kate-osborn commented on internal/state/change_processor.go:
```diff
@@ -75,40 +82,135 @@ func (c *ChangeProcessorImpl) CaptureUpsertChange(obj client.Object) {
 	c.storeChanged = c.storeChanged || resourceChanged
 }
 
+func (c *ChangeProcessorImpl) captureGatewayClassChange(gc *v1beta1.GatewayClass) bool {
```
I moved the methods to the store object.

I also moved the `storeChanged` bool from the change processor to the store, but I didn't add a `Changed()` method. Since the store and the change processor are in the same package I don't think it's necessary.

kate-osborn commented on internal/state/change_processor.go:
```diff
@@ -75,40 +82,135 @@ func (c *ChangeProcessorImpl) CaptureUpsertChange(obj client.Object) {
+		resourceChanged = false
+	}
+	c.store.httpRoutes[getNamespacedName(hr)] = hr
+	c.updateServicesMap(hr)
```
Should be fixed with new commit bda6dcf078e9dda1b5386c07dd0579f6e2610262

kate-osborn commented on internal/state/store.go:
```diff
@@ -10,11 +11,18 @@ type store struct {
 	httpRoutes map[types.NamespacedName]*v1beta1.HTTPRoute
+
+	// services maps services to the set of http routes that reference the service.
+	services map[types.NamespacedName]map[types.NamespacedName]struct{}
```
Added a relationship.Capturer to track these relationships: bda6dcf078e9dda1b5386c07dd0579f6e2610262

kate-osborn commented on internal/events/handler.go:
```diff
@@ -131,11 +121,13 @@ func (h *EventHandlerImpl) propagateUpsert(e *UpsertEvent) {
 		h.cfg.Processor.CaptureUpsertChange(r)
 	case *apiv1.Service:
-		// FIXME(pleshakov): make sure the affected hosts are updated
 		h.cfg.ServiceStore.Upsert(r)
```
Replaced ServiceStore component with a ServiceResolver that is only responsible for resolving Services to Endpoints. 

bda6dcf078e9dda1b5386c07dd0579f6e2610262

kate-osborn commented:
> @kate-osborn please see my review
> 
> A few more things:
> 
> (1) Could we update the api compatibility doc? The part about backendRefs in HTTPRoute
> 
> > NGINX Kubernetes Gateway will use the IP of the Service as a backend, not the IPs of the corresponding Pods. Watching for Service updates is not supported.
> 
> (2) I noticed a bug:
> 
> (a) deploy cafe example (b) confirm everything is deployed and working (c) change tea svc port from 80 to 7777 in the service manifest bug: config doesn’t get regenerated - keeps using the old endpoints.

Thanks @pleshakov 

I fixed the bug you found and updated the docs.

f5yacobucci commented on internal/state/configuration.go:
```diff
@@ -72,45 +85,47 @@ func buildConfiguration(graph *graph) Configuration {
+	return config
 }
 
-type configBuilder struct {
```
This doesn't appear to be a Builder or a Factory.

To be a Builder I'd expect a concrete object and an abstract builder.
```
type Builder
type Concrete

func NewBuilder() Builder // returns Builder
func (Builder) AddX() Builder // keeps returning itself as a Builder
func (Builder) AddY() Builder // keeps returning itself as a Builder
func (Builder) Build() Concrete // final stage gives us the composite object
```

Or a Factory, where one abstract method can give us multiple implementations.

I'm concerned these names diverge from their accepted patterns.

f5yacobucci replied:
I think it's getBuilderForProtocol is confusing. Shouldn't the Factory return the correct Builder by accepting arguments.

kate-osborn replied:
Valid points. I was using the name `builder` because of the intent of the package, which is to build configuration. However, I can see how that can be misleading. 

As far as the factory goes, after another look, I don't think the pattern fits the use case -- it just overcomplicates things.

I've replaced the factory and builder with what I think is a more straightforward approach. Let me know what you think.

f5yacobucci commented on internal/state/graph.go:
```diff
@@ -88,7 +123,91 @@ func buildGraph(
+				b.Endpoints, err = serviceResolver.Resolve(ctx, svc, backendSvc.Port)
+				if err != nil {
+					b.ErrorMsg = err.Error()
+					warnings.AddWarningf(r.Source, "cannot resolve backend ref for rule %d: %s", i, err.Error())
```
We print out "cannot resolve backend ref for rule..." in three (maybe more?) different branches.

The 2 here are from Resolve and the one above due to a missing backend. Is Error enough to discern what's happening? Do missing backends and an error from Resolve ulitmately have different root causes?

kate-osborn replied:
I collapsed the two resolve warnings, but I think the remaining warnings are necessary.

The warnings are a bit repetitive, but I'm trying to prevent making duplicate API calls to the k8s API server. Since rules and routes can reference the same backend service, I'm checking if the backend service has already been resolved before attempting to resolve it. 

I also think an invalid backend ref is different from a resolve failure. An invalid backend ref means the HTTPRoute config is invalid, whereas a resolve error could be a network error or a Service configuration error.

pleshakov replied:
>  I'm trying to prevent making duplicate API calls to the k8s API server.

just to double check. I don't think we make any API calls -- the Mgr client is a caching client and it should return EndpointSlices from its cache (shared informer for EndpointSlices). Do you see any cases when the API calls will still happen?  If that is case, I think we have a problem because an API call is an expensive operation.

kate-osborn replied:
You're right; it is a caching client.

f5yacobucci commented on internal/state/relationship/capturer.go:
```diff
@@ -0,0 +1,131 @@
+// The relationships between HTTPRoutes -> Services are many to 1, so these relationships are tracked using a counter. A Service relationship exists if at least one HTTPRoute references it.
+// An EndpointSlice relationship exists, if its Service owner is referenced by at least one HTTPRoute.
+type Capturer struct {
+	routesToServices    map[types.NamespacedName]map[types.NamespacedName]struct{}
```
Type declarations would add to readability.

kate-osborn replied:
Added

f5yacobucci commented on internal/state/relationship/capturer.go:
```diff
@@ -0,0 +1,131 @@
+	delete(c.routesToServices, routeName)
+}
+
+func (c *Capturer) decrementRefCount(svcName types.NamespacedName) {
```
This function can leave us with a key and -1.

```
package main

import "fmt"

func dec(m map[string]int, key string) {
	if m[key] == 1 {
		delete(m, key)
		return
	}
	m[key]--
}

func main() {
	m := make(map[string]int)

	dec(m, "hello")
	fmt.Printf("%+v\n", m)
}
```

f5yacobucci replied:
`map[hello:-1]`

f5yacobucci replied:
We may need a test that exercises key existence versus non-existence.

kate-osborn replied:
Thanks for catching this.

I fixed it and am now explicitly testing the ref count.

f5yacobucci commented on internal/state/relationship/capturer.go:
```diff
@@ -0,0 +1,131 @@
+// Currently, it only captures relationships between HTTPRoutes and Services and Services and EndpointSlices, but it can be extended to capture additional relationships.
+// The relationships between HTTPRoutes -> Services are many to 1, so these relationships are tracked using a counter. A Service relationship exists if at least one HTTPRoute references it.
+// An EndpointSlice relationship exists, if its Service owner is referenced by at least one HTTPRoute.
+type Capturer struct {
```
nit: idiomatically an '-er' ending means an interface. Is there another name we feel comfortable with?

Captures? Affinities? Edges? SvcDependencies? Appositives (j/k)...I don't have a great alternative.

pleshakov replied:
CapturerImpl ;)

kate-osborn replied:
Changed to `CapturerImpl` to be consistent with the rest of the code base.

f5yacobucci commented on internal/state/resolver/resolver.go:
```diff
@@ -0,0 +1,153 @@
+
+func targetPortExists(ports []discoveryV1.EndpointPort, targetPort int32) bool {
+	for _, port := range ports {
+		if port.Port == nil {
```
"When a port is defined with a nil port value, it indicates "all ports"."

Does anyone know what Service definitions would cause an "all ports" endpoint configuration? I read all ports as indicating either this should return `true` or we test and document special cases where we don't support a Service config.

kate-osborn replied:
The only way I can produce an EndpointSlice with nil ports is by creating a Service that has no endpoints:
```
addressType: IPv4
apiVersion: discovery.k8s.io/v1
endpoints: null
kind: EndpointSlice
metadata:
  creationTimestamp: "2022-09-20T22:51:50Z"
  generateName: target-svc-
  generation: 1
  labels:
    endpointslice.kubernetes.io/managed-by: endpointslice-controller.k8s.io
    kubernetes.io/service-name: target-svc
  name: target-svc-pxhxr
  namespace: default
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    controller: true
    kind: Service
    name: target-svc
    uid: 903364aa-a01e-4087-b72b-f0229cf2f84b
  resourceVersion: "2897033"
  uid: 13571449-21d2-4553-8be1-774829204869
ports: null
```

kate-osborn replied:
nvm that's a different case -- `When ports is empty, it indicates that there are no defined ports.`

kate-osborn replied:
I'm not sure what Service definition would cause an "all ports" endpoint config. 

`Service.Spec.Ports` is a required field as is `Service.Spec.Ports.Port`

kate-osborn replied:
> I read all ports as indicating either this should return true 

Agreed. I'll fix this to return true.

kate-osborn replied:
Asked about this in the Kubernetes slack: https://kubernetes.slack.com/archives/C09NXKJKA/p1663971959776179

kate-osborn replied:
Haven't heard back from the k8s owners, but after reading the endpoint slice code it looks like the only way this can happen is if a user manually creates an endpoint slice object.

brianehlert replied:
`By default, the control plane creates and manages EndpointSlices to have no more than 100 endpoints each. You can configure this with the --max-endpoints-per-slice [kube-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) flag, up to a maximum of 1000.`
https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/

pleshakov commented on internal/state/relationship/capturer_test.go:
```diff
@@ -0,0 +1,304 @@
+			capturer = relationship.NewCapturer()
+		})
+		Describe("Normal cases", Ordered, func() {
+			When("a route with a backend service is captured", func() {
```
using when makes it much more readable 👍  I'll make sure to use it in my future tests

pleshakov commented on internal/state/resolver/service_resolver_test.go:
```diff
@@ -0,0 +1,180 @@
+	Describe("Resolve", Ordered, func() {
+		BeforeAll(func() {
+			var err error
+			fakeK8sClient, err = createFakeK8sClient(slice1, slice2, sliceIPV6, slicePort8081)
```
would it make sense to add a comment that "fakeK8sClient will return all EndpointSlices for the `List()` call done by the ServiceResolver"? 
Basically, the fact that the resolver uses `Service Name Index Field ` doesn't matter.

kate-osborn commented on internal/state/relationship/capturer.go:
```diff
@@ -0,0 +1,131 @@
+// Currently, it only captures relationships between HTTPRoutes and Services and Services and EndpointSlices, but it can be extended to capture additional relationships.
+// The relationships between HTTPRoutes -> Services are many to 1, so these relationships are tracked using a counter. A Service relationship exists if at least one HTTPRoute references it.
+// An EndpointSlice relationship exists, if its Service owner is referenced by at least one HTTPRoute.
+type Capturer struct {
```
I struggled a bit with this component but hoped to make it extensible for resources like Secrets.

When we support certificate rotation, we will need to process secret updates, and only trigger an update when a secret is referenced by an existing listener.

This component could be extended to track the listener -> secret relationship.

I am tracking the EndpointSlice owners in this component to handle the EndpointSlice delete event. On a delete event, we do not have access to the EndpointSlice object and cannot look up the service owner.

pleshakov replied:
thanks for clarifying! this makes sense to me now

kate-osborn commented on pkg/sdk/endpointslice_controller.go:
```diff
@@ -0,0 +1,102 @@
+		impl:   impl,
+	}
+
+	err := mgr.GetFieldIndexer().IndexField(context.TODO(), &discoveryV1.EndpointSlice{}, KubernetesServiceNameIndexField, ServiceNameIndexFunc)
```
No, I should just fix it now. I keep using `context.TODO()` as a placeholder and then forgetting to replace it. 

Do you think `context.WithTimeout` is reasonable here?

pleshakov replied:
context.WithTimeout sounds reasonable. However, I think it should also depend on the signal handler context, so that if Gateway pod gets shutdown, the IndexField unblocks.
```
	ctx := ctlr.SetupSignalHandler()
```

kate-osborn replied:
Ah, yes, good idea. Do you think using the ctx from the controller is sufficient?

pleshakov replied:
> Do you think using the ctx from the controller is sufficient?

If we use that ctx, will there be a chance that the IndexField() hangs indefinitely? And, as a result, the pod will just hang? If that could happen, I think a timeout based on the controller ctx is a good idea.

kate-osborn commented on internal/state/change_processor.go:
```diff
@@ -71,77 +78,86 @@ func (c *ChangeProcessorImpl) CaptureUpsertChange(obj client.Object) {
 	}
 
-	c.storeChanged = c.storeChanged || resourceChanged
+	c.store.changed = c.store.changed || c.relationshipCapturer.Exists(obj, client.ObjectKeyFromObject(obj))
```
I can introduce a new boolean for the ChangeProcessor. Still, I don't think we can entirely remove its dependency on the store because we need a way to reset the `store.changed` variable after a Process event.

Currently, we do this in `ChangeProcessor.Process()`: https://github.com/nginxinc/nginx-kubernetes-gateway/blob/3c90c2296b5c1caf600f7e6a5ff8b766cd9a0a2c/internal/state/change_processor.go#L136.

pleshakov replied:
> c.storeChanged = false 

That looks good to me. In that case, we just reseting the store. However, `c.store.changed = c.store.changed || c.relationshipCapturer.Exists(obj, client.ObjectKeyFromObject(obj))`, we're adding additional meaning to `c.store.changed`, so that it also reflects changed relationship between objects, even if they are not present in the store.

f5yacobucci commented on internal/state/relationship/capturer.go:
```diff
@@ -0,0 +1,131 @@
+	"github.com/nginxinc/nginx-kubernetes-gateway/pkg/sdk"
+)
+
+// Capturer captures relationships between Kubernetes objects and can be queried for whether a relationship exists for a given object.
```
There is (or at least used to be) a linter to limit line length. Can we enable that?

kate-osborn replied:
https://golangci-lint.run/usage/linters/#lll

yep, I'll enable it.

kate-osborn replied:
Actually, I'll add this in a separate PR. There are tons of changes that need to be made.

pleshakov commented on internal/state/relationship/capturer.go:
```diff
@@ -95,8 +123,8 @@ func (c *Capturer) deleteForRoute(routeName types.NamespacedName) {
-func (c *Capturer) decrementRefCount(svcName types.NamespacedName) {
-	if c.serviceRefCount[svcName] == 1 {
+func (c *CapturerImpl) decrementRefCount(svcName types.NamespacedName) {
+	if c.serviceRefCount[svcName] == 0 {
```
because we remove only when it is 0 (not when 1 as it was before), can we have a case when a lot of no longer referenced services linger in the map? 

Ex:
a HTTPRoute that refs svcA is created  -> serviceRefCount is {'ns/A': 1}
the HTTPRoute is updated to ref svc B -> serviceRefCount is  {'ns/A': 0, 'ns/B': 1}

Will 'ns/A' ever get removed?

kate-osborn replied:
I can update this to check if ref count is 0 || 1. That should handle both cases.

kate-osborn replied:
https://github.com/nginxinc/nginx-kubernetes-gateway/pull/221/commits/59f8c711a25dedc95b14cf93c7a60dd25b1a758a

pleshakov commented on internal/state/resolver/resolver.go:
```diff
@@ -102,7 +106,7 @@ func getTargetPort(svc *v1.Service, svcPort int32) (int32, error) {
 		if port.Port == svcPort {
 			val := port.TargetPort.IntValue()
```
the target port can be a string like "myport" rather than "8080". In this case, the actual port number can looked up in the pod spec like here https://github.com/nginxinc/kubernetes-ingress/blob/b69bc90cd6d5603f34744780cc5b4ace038c93fc/internal/k8s/utils.go#L105 

I think we should support it but perhaps we can leave it as a FIXME for now?  

sorry, didn't catch that earlier.

kate-osborn replied:
Thanks for catching. I overlooked this case. 

I can fix it now by looking up the actual port number in the EndpointSlice ports list.

kate-osborn replied:
Added logic for string target ports: https://github.com/nginxinc/nginx-kubernetes-gateway/pull/221/commits/ccbebd6b1d7e01c58997933ac14082baa95362e1

kate-osborn commented on internal/state/graph.go:
```diff
@@ -161,17 +161,11 @@ func resolveBackends(
-					warnings.AddWarningf(r.Source, "cannot resolve backend ref for rule %d: %s", i, err.Error())
-				}
-			} else {
-				b.ErrorMsg = fmt.Sprintf("Service %s does not exist", svcName)
```
I agree. I was trying to collapse the number of error messages. 

Let me try and rework this section again.

kate-osborn replied:
I added back in the Service does not exist error.

https://github.com/nginxinc/nginx-kubernetes-gateway/pull/221/commits/dcfae30be46945e46b874722f196ebf7d26de51e

pleshakov commented on internal/state/resolver/resolver.go:
```diff
@@ -58,41 +59,42 @@ func (e *ServiceResolverImpl) Resolve(ctx context.Context, svc *v1.Service, svcP
 	}
 
-	capacity := calculateEndpointSliceCapacity(endpointSliceList.Items, targetPort)
+	filteredSlices := filterEndpointSliceList(endpointSliceList, svcPort)
```
To calculate the capacity, we filter the list, which allocates `filtered := make([]discoveryV1.EndpointSlice, 0, len(endpointSliceList.Items))`. 

Considering that, would it be easier (and same in terms of memory)  if we just `endpoints := make([]Endpoint, 0, len(endpointSliceList.Items))`

kate-osborn replied:
There can be multiple endpoints in an EndpointSlice. `endpointSliceList.Items` is just the list of EndpointSlices.

pleshakov replied:
my bad. misread the code :(

pleshakov commented on internal/state/resolver/resolver.go:
```diff
@@ -142,13 +149,53 @@ func endpointReady(endpoint discoveryV1.Endpoint) bool {
+// ServicePort is an int, we find the endpoint port by value. If the endpoint port is nil, then according to the spec,
+// all ports should be considered a match. In this case, we return the ServicePort port value for string targetPorts,
+// or the targetPort integer value. If no match is found, return 0.
+func findPort(ports []discoveryV1.EndpointPort, svcPort v1.ServicePort) int32 {
```
I caught the following bug:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: coffee
spec:
  replicas: 1
  selector:
    matchLabels:
      app: coffee
  template:
    metadata:
      labels:
        app: coffee
    spec:
      containers:
      - name: coffee
        image: nginxdemos/nginx-hello:plain-text
        ports:
        - containerPort: 8080
          name: myport
---
apiVersion: v1
kind: Service
metadata:
  name: coffee
spec:
  ports:
  - port: 80
    targetPort: myport
    protocol: TCP
    name: http
  selector:
    app: coffee
```    

EndpointSlice ports:
```yaml
ports:
- name: http
  port: 8080
  protocol: TCP
```

logs:
```
{"level":"info","ts":1664299471.3182275,"logger":"changeProcessor","msg":"Got warning while building graph","kind":"HTTPRoute","namespace":"default","name":"coffee","warning":"cannot resolve backend ref for rule 0: no valid endpoints found for Service default/coffee and port {http TCP <nil> 80 {1 0 myport} 0}"}
```

expected - no error.

I also wonder if it make sense to print the name of the fields for the port, so that it would be easier to parse it for humans in the logs?

kate-osborn replied:
🤦‍♀️ I should've tested with a service port name that was different from the target port name. Thanks for catching. 

So the k8s EndpointSlice controller is already doing the mapping of target port to container port. I just need to match the service port name to the endpoint port name.

However, the service port name is optional if there's only one service port defined. 

I need to do more testing.

kate-osborn replied:
Confirmed that the Kubernetes EndpointSlice controller is handling all of the logic of mapping target port to container port (see this [function](https://github.com/kubernetes/kubernetes/blob/5835544ca568b757a8ecae5c153f317e5736700e/pkg/controller/endpointslice/utils.go#L85)).

I updated `findPort` to return the EndpointPort port that matches the ServicePort name. In the case where the ServicePort is unnamed, we can still use this logic because the corresponding EndpointPort port will also be unnamed (unnamed is defined as a name with an empty string value).

pleshakov commented on internal/state/resolver/service_resolver_test.go:
```diff
@@ -99,22 +101,25 @@ var _ = Describe("ServiceResolver", func() {
 						Port: 443,
 						TargetPort: intstr.IntOrString{
 							Type:   intstr.String,
 							StrVal: "8443",
```
I think `StrVal: "8443"` can be misleading, because k8s will set the int value if YAML includes "8443", not string.

pleshakov reviewed (approved):
👍
